implemented issue#917

zxkmm opened the pull request:
I also noticed issue#916, but i did a test and it seems there's no much more memory were taken even if i decleared a new variable....   
When rename a file, just cut the extension from name_buffer, and when the keypad return the buffer, just add the extension back so user wouldn't need to delete the extension and type it again....   
it is implemented in "refactor" button so user can still rename the extension with "rename" if they need.    
considering record and replay is the the main feature of pp so i think it worth to waste SPI flash memory within 11 line code to implement it...

gullradriel commented:
Looks good. Have some tested it?

zxkmm commented:
Tested all the situation I can imagine and nothing wrong.

zxkmm commented:
Ohhhh don’t merge…. There’s still a bug

zxkmm commented:
it seems ``clear()`` sometimes work, sometimes not.... I don't know why..... maybe ill close this later.

gullradriel commented:
std::string clear: Erases the contents of the string, which becomes an empty string with a length of 0.
string is not destroyed, it's cleared out.
if you still have data inside, then you have not passed y the clear function (maybe a test is not matching, or returning before calling clear)

zxkmm commented:
I didn't want it destroyed. it works all good without it. i added it in second commit is because thing getting random (even files getting deleted!) after I keeping renaming after about 5 or 6 times. i think I leaked memory somehow with ``extension_buffer`` so i added clear() but it doesn't getting better.

gullradriel commented on firmware/application/apps/ui_fileman.cpp:
```diff
@@ -267,14 +267,15 @@ void FileManagerView::on_refactor(NavigationView& nav) {
 			destination_path += '/';
-		destination_path = destination_path + buffer;
+
+		destination_path = get_selected_path().string().back() != '/' ? destination_path + buffer + extension_buffer : destination_path + buffer;
```
unclear and complicated ternary operation. use a if else like and clarify what you're doing here.

zxkmm replied:
ty, I can but from what i saw ``?:`` is same. and what about files getting deleted?

gullradriel replied:
Something like this i.e :

if( get_selected_path().string().back() != '/' )  
{
    //adding back extension since it's not a directory
    destination_path += buffer + extension_buffer ;
}
else  
{
    destination_path += buffer ;
}

gullradriel replied:
git killed my indentation but you got the idea.

zxkmm replied:
updated.. bug still there, after i constantly renaming 5 or 6 files, files getting deleted and the extension_buffer becoming empty no matter what i do.

gullradriel replied:
> ty, I can but from what i saw `?:` is same. and what about files getting deleted?
It's less readable and not quicker, maybe only IMHO.
I do not get the file deleted thing, you have to give more details.

zxkmm replied:
>I do not get the file deleted thing, you have to give more details.

I'm renaming files, all good, but at some point, the file I'm renaming, getting deleted. it shouldn't happen.

zxkmm replied:
>ho you mean the content of the var is dead ?
no.
the real file at the sdcard getting deleted...
Sounds weird but it happened.

gullradriel commented:
okay I think file can be deleted because of a bad rename

gullradriel commented:
rename_file(get_selected_path(), destination_path);  //rename the selected file

auto selected_path = get_selected_path();  ==> I don't think get_selected_path point to a good value anymore, shouldn't it be destination_path ?
auto extension = selected_path.extension().string();

gullradriel commented on firmware/application/apps/ui_fileman.cpp:
```diff
@@ -267,14 +267,15 @@ void FileManagerView::on_refactor(NavigationView& nav) {
 
 		rename_file(get_selected_path(), destination_path);  //rename the selected file
 
 		auto selected_path = get_selected_path();
```
shouldn't we use destination_path here since we renamed, but not reloaded via load_directory_content yet ?

zxkmm replied:
> shouldn't we use destination_path here since we renamed, but not reloaded via load_directory_content yet ?

yes, im so sorry....

zxkmm commented:
> ``auto selected_path = get_selected_path();`` ==> I don't think get_selected_path point to a good value anymore, shouldn't it be ``destination_path`` ?

you are right, im' so sorry. testing.

gullradriel commented:
I think it's a bit more than just that

gullradriel commented:
Looking at get_selected_path we should just update current_path with destination_path before calling get_selected_path

zxkmm commented:
have it fixed.
```
		if(extension.empty()){// Is Dir
			destination_path = destination_path + buffer;
			extension_buffer = "";
		}else{//is File
			destination_path = destination_path + buffer + extension_buffer;
		}


```

gullradriel commented:
Great, now test more and fix if needed ^^

zxkmm commented:
Tested possible situation: 
``test.TXT`` and ``test.C16``  Behave normal    
``testfile.abc``                   Behave normal    
``test.dot.txt``                   Behave normal  
``test space.txt``               Behave normal  
``testdir/`` right after ``testfile.abc``   Behave normal    
``test sapce dir/``               Behave normal    
``test.dot.dir/``                      can't rename things after last dot, but would not crash and can still rename it using "Rename", will try to fix it tho.

Filename^         if_normal^

zxkmm commented:
> Somewhere they use if (std::filesystem::is_directory(entry.status())
Maybe try your luck with it ?

that would be tricky, `std::filesystem::is_directory()` is for `std::filesystem::file_status` but `get_selected_path()` returns a `std::filesystem::path` . idk how to convert and compare.

zxkmm commented:
thx again for such a spoon feeding level of helping.  
i felt guilty for spamming such useless PR. 
test vid ->
https://youtu.be/uKa6l87lInw
